### PR TITLE
Associate weight with p [For push-sum algorithm]

### PR DESCRIPTION
**Pull request by Bluefog-Lib**

### CODE REVIEW

**Review by hanbinhu — CHANGES_REQUESTED** [review]

File: bluefog/torch/mpi_ops.py [hunk 1]
```diff
@@ -1053,3 +1069,15 @@ def _win_mutex_acquire(name, ranks, exclusive):
 
 def _win_mutex_release(name, ranks, exclusive):
     mpi_lib.bluefog_torch_win_mutex_release(name, ranks, exclusive)
+
+
+def win_associated_p(name: str) -> float:
```
[the remark this answers is not in the document]
Comment (Bluefog-Lib) [reply bluefog/torch/mpi_ops.py:1074]: Done

File: bluefog/torch/mpi_ops.py [hunk 2]
```diff
@@ -1053,3 +1069,15 @@ def _win_mutex_acquire(name, ranks, exclusive):
...
+    return mpi_lib.bluefog_torch_win_associated_p(name)
+
+
+def turn_on_win_ops_with_associated_p():
```
[the remark this answers is not in the document]
Comment (Bluefog-Lib) [reply bluefog/torch/mpi_ops.py:1078]: Done

File: bluefog/torch/mpi_ops.py [hunk 3]
```diff
@@ -1053,3 +1069,15 @@ def _win_mutex_acquire(name, ranks, exclusive):
...
+    mpi_lib.bluefog_torch_set_win_ops_with_associated_p_state(True)
+
+
+def turn_off_win_ops_with_associated_p():
```
[the remark this answers is not in the document]
Comment (Bluefog-Lib) [reply bluefog/torch/mpi_ops.py:1082]: Done

File: test/torch_win_ops_test.py [hunk 4]
```diff
@@ -629,6 +627,84 @@ def test_win_mutex_given_ranks(self):
...
+        rank = bf.rank()
+        if size <= 3:
+            fname = inspect.currentframe().f_code.co_name
+            warnings.warn(
+                "Skip {} because it only supports test over at least 3 nodes".format(fname))
+            return
+
+        dtypes = [torch.FloatTensor, torch.DoubleTensor]
+        if TEST_ON_GPU:
+            dtypes += [torch.cuda.FloatTensor, torch.cuda.DoubleTensor]
```
[the remark this answers is not in the document]
Comment (Bluefog-Lib) [reply test/torch_win_ops_test.py:641]: So far. Yes. I don't see the problem yet.

File: test/torch_win_ops_test.py [hunk 5]
```diff
@@ -629,6 +627,84 @@ def test_win_mutex_given_ranks(self):
...
+                               for i, r in enumerate(bf.out_neighbor_ranks())}
+                bf.win_accumulate(tensor, name=window_name, require_mutex=True,
+                                  self_weight=self_weight, dst_weights=dst_weights)
+                bf.win_update_then_collect(name=window_name)
+            bf.barrier()
+            bf.win_update_then_collect(name=window_name)
+            associated_p = bf.win_associated_p(name=window_name)
+            # Because the associated p should operate the same as tensor always
+            # the following assert should be true no matter what order is excuted.
+            assert abs(associated_p - tensor.data[0])< 0.1
```
[the remark this answers is not in the document]
Comment (Bluefog-Lib) [reply test/torch_win_ops_test.py:704]: You are right. Should use epsilon

**Pull request merged**